Малямов Иван

ivanarray opened the pull request:
review: @skylog
#TagCloud

ivanarray committed:
Начал реализацию задачи.
review: @skylog
#TagCloud

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,51 @@
+            return result;
+        }
+
+        public void PutNextRectangles(params Size[] sizes)
```
используется ведь только в тестах, точно нужен этот метод?

ivanarray replied:
> используется ведь только в тестах, почему решил так реализовать? 
Для тестов, если это критично могу убрать в тестовый класс.

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+{
+    [TestFixture]
+    public class Tests
+    {
```
Давай приведем неймы методов к формату MethodName_When_ExpectedBehavior, пример: IsAdult_AgeLessThan18_False

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+namespace TagsCloudVisualization
+{
+    [TestFixture]
+    public class Tests
```
Нужно дать более описательное имя

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+            var cloud = new CircularCloudLayouter(center);
+
+            var size = new Size(4, 4);
+            var x = center.X - 2;
```
а для чего center.X - 2 и center.Y + 2 ?

ivanarray replied:
![image](https://user-images.githubusercontent.com/73218360/140766028-0bd20330-8980-45a9-b672-ef4c9c8775d5.png)
Нам нужно вычислить Rectangle.location, Centre- центр нашего облака. Я понял намек, что нужно писать понятные тесты)) и у меня похоже тут ошибка

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+        {
+            var cloud = new CircularCloudLayouter(center);
+
+            var size = new Size(4, 4);
```
почему именно такой Size(4, 4) ?

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+            var y = center.Y + 2;
+            var expectedLocation = new Point(x, y);
+
+            var actual = cloud.PutNextRectangle(size);
```
var actual - нужно более описательное имя

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+        [Test]
+        public void AfterPutNextRectangle_should_not_any_intersecting_rectangles()
+        {
+            var cloud = new CircularCloudLayouter();
```
var cloud  - нужно более описательное имя

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+
+            cloud.PutNextRectangles(sizes);
+
+            var intersectingRectanglePairs =
```
- это сильно усложняет тест, и делает его более хрупким, тут лучше подобрать граничные значения для проверки через TestCase

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,51 @@
+                PutNextRectangle(s);
+        }
+
+        private Point GetRectangleLocation(Size rectangleSize)
```
Подумай над ответственностью CircularCloudLayouter, и не нарушает ли метод GetRectangleLocation single-responsibility principle?

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,51 @@
+            return new Point(0, 0);
+        }
+
+        public IEnumerable<Rectangle> GetRectangles()
```
тут будет лучше не IEnumerable, если нужно вернуть коллекцию только для чтения советую глянуть - https://enterprisecraftsmanship.com/posts/ienumerable-vs-ireadonlylist/

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,51 @@
+        private readonly List<Rectangle> rectangles = new();
+
+
+        public CircularCloudLayouter(Point center = new())
```
давай не будем тут использовать дефолтное значение? это будет работать только потому что Point структура, и подумай про ответственность CircularCloudLayouter

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,51 @@
+
+        private Point GetRectangleLocation(Size rectangleSize)
+        {
+            if (rectangles.Count == 0)
```
а тут точно нет ошибки?

ivanarray replied:
В названии?

skylog replied:
`if (rectangles.Count == 0)` - это временная заглушка? (только первый прямоугольник положим в центр)

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+        };
+
+        [TestCaseSource(nameof(centralPoints))]
+        public void Constructor_should_initialize_without_exceptions(Point centre)
```
в конструкторе CircularCloudLayouter нет логики проброса исключения, тест лукавит, ведь тут тестируется Point, уверен что нужно тестировать Point?

skylog commented on cs/TagsCloudVisualization/Tests.cs:
```diff
@@ -0,0 +1,93 @@
+                "Location can to have positive and negative coordinates");
+        }
+
+        [TestCaseSource(nameof(centralPoints))]
```
тут TestCaseSource усложняет чтение теста, посмотри когда хорошо использовать TestCaseSource а когда TestCase

skylog reviewed (commented):
предпромежуточное ревью

ivanarray committed:
Добавил класс GeometryHelper.cs для некоторых геометрических вычислений и тесты к нему.
review: @skylog
#TagCloud

ivanarray committed:
Изменил тесты, добавил тест для простого случая.
review: @skylog
#TagCloud

ivanarray reviewed (commented):
>Изменил тесты, добавил тест для простого случая.

Для простого случая добавления прямоугольников*

skylog commented:
*в тестах предложил бы почитить и воспользоваться [SetUp], чтобы инициализировать дефолтное расположение, CircularCloudLayouter, и может что еще

skylog commented on cs/TagsCloudVisualization/GeometryHelper.cs:
```diff
@@ -0,0 +1,32 @@
+            var x = rect.X + rect.Width / 2.0f;
+            var y = rect.Y + rect.Height / 2.0f;
+
+            return new PointF(x, y);
```
странно когда используется Point и PointF, можешь рассказать почему так?

ivanarray committed:
Реализовал начальный вариант алгоритма поиска локации, нужно дорабатывать
review: @skylog
#TagCloud

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -27,13 +38,49 @@ public Rectangle PutNextRectangle(Size rectangleSize)
-            if (rectangles.Count == 0)
-                return GeometryHelper
-                    .GetRectangleLocationFromCentre(Center, rectangleSize);
+            while (currentRadius < limiter)
```
а есть ограничение для первого прямоугольника, его стороны не могут быть > 100?

skylog replied:
эту логику нужно вынести в stepProperties, например в bool метод IsValidRadius()

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -1,14 +1,25 @@
+        private float currentRadius;
+        private float radiusStep = 1;
+        private int currentStep = 1;
+        private int limiter = 100;
```
эта константа limiter не очень понятна для чего нужна

ivanarray committed:
Сделал класс Painter.cs для отрисовки результата, так-же добавил метод в хелпер для определения размера изображения

review: @skylog
#TagCloud

ivanarray committed:
Добавил tearDown метод, для отрисовки неудачного результата.
Так-же написал тест, где добавляются прямоугольники разных размеров, и он валится)) допиливаю алгоритм

review: @skylog
#TagCloud

ivanarray committed:
Сделал рабочую версию
review: @skylog
#TagCloud

ivanarray commented:
Запушил вообще без рефакторинга, сейчас этим занимаюсь

ivanarray committed:
Добавил картинки

ivanarray committed:
Преобразовал bmp в jpg, добавил README.md

ivanarray committed:
Преобразовал bmp в jpg

ivanarray committed:
Merge remote-tracking branch 'origin/master'

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+
+            currentRadius = Math.Min(rectangleSize.Height, rectangleSize.Width);
+            return GeometryHelper
+                .GetRectangleLocationFromCentre(Center, rectangleSize);
```
.GetRectangleLocationFrom**Centre**(**Center**, rectangleSize);
Приведи пожалуйста к единому неймингу, либо как в UK - Centre, либо как в USA - Center, это касается всего решения

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+        private readonly HashSet<Point> usedPoints = new();
+
+
+        private float currentStepAngle = MathF.PI / 2;
```
нужно выкинуть отсюда все настройки в отдельный класс StepProperties, тут  в свойствах/полях должны остаться только 
![image](https://user-images.githubusercontent.com/7670395/142026373-030a9bb5-87ac-4bb5-9fc9-4de389171e62.png)

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+            {
+                var angle = currentStep * currentStepAngle;
+
+                while (angle < Pi2)
```
эту логику нужно вынести в stepProperties, например в bool метод IsValidAngle()

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            while (currentRadius < limiter)
+            {
+                var angle = currentStep * currentStepAngle;
```
нужно в конструкторе stepProperties задать дефолтное значение angle, а обновлять внутри цикла а не здесь

skylog replied:
в конструкторе stepProperties можно сразу задать дефолтные значения для step, count, angle

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+                while (angle < Pi2)
+                {
+                    var rectangleCentre =
+                        GeometryHelper.GetPointOnCircle(Center, currentRadius, angle);
```
в stepProperties будет св-ва Angle и Radius

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+                        GeometryHelper.GetPointOnCircle(Center, currentRadius, angle);
+                    var rectangleLocation =
+                        GeometryHelper.GetRectangleLocationFromCentre(rectangleCentre, rectangleSize);
+                    angle = currentStep * currentStepAngle;
```
внутри геттера св-ва можно заинлайнить эту логику ` Angle { get => angle *= Step}`

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+
+        private void UpdateParameters()
+        {
+            currentRadius += radiusStep;
```
внутри геттера св-ва можно заинлайнить эту логику ` Radius {get => Radius += RadiusStep;  }`

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+        }
+
+
+        private void UpdateParameters()
```
нужно перенести в stepProperties, а дать какое-нибудь понятное имя

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            if (rectangles.Count != 0) return SearchPointOnCircle(rectangleSize);
+
+            currentRadius = Math.Min(rectangleSize.Height, rectangleSize.Width);
```
в stepProperties можно завести метод SetInitialRadius чтобы задать начальный Radius

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+
+
+                    if (!rectangles.Any(r => r.IntersectsWith(rectangle)))
+                    {
```
в stepProperties можно перенести эту логику в метод IncrementStep, ну или с более подходящим неймом

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,103 @@
+                        currentStep = 1;
+                        currentRadius = 1;
+                        currentStepAngle = MathF.PI / 2;
+                        return rectangle.Location;
```
и в IncrementStep не будет return rectangle.Location, а только обновление параметров (только имя метода подбери подходящее)

ivanarray committed:
Небольшой рефакторинг

ivanarray committed:
Отрефакторил. Выделил класс CloudLayouterParameters.cs, добавил тесты к нему
review: @skylog
#TagCloud

ivanarray committed:
Обновил README.md

ivanarray committed:
Update README.md

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,79 @@
+
+        private readonly List<Rectangle> rectangles = new();
+        private readonly HashSet<Point> usedPoints = new();
+        public readonly CloudLayouterParameters Parameters = new();
```
- св-во должнобыть private
- lowercase

skylog replied:
нейм класса не должен быть не CloudLayout**er**Parameters, а "параметры раскладки", ну и само св-во можно не сокращать до Parameters

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,79 @@
+
+        public Rectangle PutNextRectangle(Size rectangleSize)
+        {
+            if (rectangleSize.Height < 0 || rectangleSize.Width < 0)
```
по решению не хватает форматирования кода, хотелось бы видеть фигурные скобки в блоках if
` if(conditions)`
`{
DoSomething()
}
 `

skylog commented on cs/TagsCloudVisualization/GeometryHelper.cs:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        public static Size GetRectanglesCommonSize(IEnumerable<Rectangle> rectangles, int framing = 500)
+            => rectangles.Aggregate(new Rectangle(), Rectangle.Union).Size
```
- сложно читать такие однострочные методы, лучше вышести в переменную операцию агрегации `rectangles.Aggregate(new Rectangle(), Rectangle.Union)`
- вместо new Rectangle можно воспользоваться Rectangle.Empty

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,79 @@
+                return SearchRectangleOnCircle(rectangleSize);
+
+            var rectangleLocation = GeometryHelper
+                .GetRectangleLocationFromCenter(Center, rectangleSize);
```
переносы тоже не особо добавляют к читабельности, это и к остальным вызовам методов GeometryHelper

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,79 @@
+        {
+            while (true)
+            {
+                while (Parameters.IsNextAngleLessThanPi2())
```
IsNextAngleLessThanPi2 - такое имя раскрывает детали реализации (при изменении реализации придется ренеймить метод), тут должно быть что-то более абстрактное

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,79 @@
+            {
+                while (Parameters.IsNextAngleLessThanPi2())
+                {
+                    var rectangleCenter =
```
читабельность кода можно легко улучшить вынеся код c 55 по 61 строку в отдельный метод HasIntersectPoints ( тут ведь проверяем пересечение с точками?)
```
if (HasIntersectPoints(rectangleSize, out var rectangleLocation))
                    {
                        continue;
                    }
```

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,79 @@
+
+                    if (usedPoints.Contains(rectangleLocation)) continue;
+
+                    var rectangle = new Rectangle(rectangleLocation, rectangleSize);
```
читабельность кода можно легко улучшить вынеся код c 63 по 65 строку в отдельный метод HasIntersectRectangle
`if (HasIntersectRectangle(rectangleSize, rectangleLocation, out var rectangle))
                    {
                        continue;
                    }`

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+            MinAngle = minAngle;
+        }
+
+        public int MaxRadius { get; }
```
с 24 по 28 строки должен быть private

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+    {
+        private const float Pi2 = MathF.PI * 2;
+
+        public CloudLayouterParameters(
```
оставь в параметрах конструктора только те параменты что задаешь (меняешь с дефолтных)

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public CloudLayouterParameters(
+            float startAngle = MathF.PI / 2,
+            float minAngle = MathF.PI / 180 * 5,
```
нужно вынести в константу

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+        public CloudLayouterParameters(
+            float startAngle = MathF.PI / 2,
+            float minAngle = MathF.PI / 180 * 5,
+            int stepLength = 10,
```
нужно вынести в константу

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+            float minAngle = MathF.PI / 180 * 5,
+            int stepLength = 10,
+            int countRadiusBoostsForChangeAngle = 5,
+            int maxRadius = 10000
```
нужно вынести в константу

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+        {
+            StepAngle = StartAngle = startAngle;
+            MaxRadius = maxRadius;
+            CurrentRadius = StepLength = stepLength;
```
получилось очень много свойств, давай избавимся от префиксов Current

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public int MaxRadius { get; }
+        public int CountRadiusBoostsForChangeAngle { get; }
+        public int StepLength { get; }
```
вороде и без Length понятно

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+
+namespace TagsCloudVisualization
+{
+    public class CloudLayouterParameters
```
Попробуй упростить, тут получилось слишком много параметров, думал что у тут будут только публичные св-ва: Step, Count, Angle, 
и приватные константы Pi2 = 2 * MathF.PI; Limiter = 100000; RadiusStep = 10; ну и методы.

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+        public float StepAngle { get; private set; }
+
+        
+        private int CurrentStep { get; set; }
```
а зачем тут приватное св-во?, тут хватило бы и поля

skylog commented on cs/TagsCloudVisualization/CloudLayouterParameters.cs:
```diff
@@ -0,0 +1,68 @@
+            StepAngle = StartAngle = startAngle;
+            MaxRadius = maxRadius;
+            CurrentRadius = StepLength = stepLength;
+            CountRadiusBoostsForChangeAngle = countRadiusBoostsForChangeAngle;
```
очень сложный нейм, может можно сделать проще?

skylog commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouter_Tests.cs:
```diff
@@ -0,0 +1,130 @@
+        [TestCase(0, 0, 2, 2, -1, -1)]
+        [TestCase(0, 0, 3, 3, -2, -2)]
+        [TestCase(0, 0, 0, 0, 0, 0)]
+        [TestCase(3, 4, 5, 7, 0, 0)]
```
Может кейс когда координаты центра cX, cY != 0 вынести в отдельный тест? Так получится сократить кол-во параметров метода.

skylog commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouter_Tests.cs:
```diff
@@ -0,0 +1,130 @@
+        {
+            var size = new Size(3, 2);
+
+            PutManiRectangles(size, 20);
```
опечатка в Mani

skylog commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouter_Tests.cs:
```diff
@@ -0,0 +1,130 @@
+        }
+
+
+        private bool AnyRectanglesIsIntersecting(IEnumerable<Rectangle> rectangles)
```
не нужно выносить названия linq методов в названия методов (от этого не станет понятней), + тесты могут читать не только разработчики, понятней будет назвать метод - HasIntersectedRectangles

skylog commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouter_Tests.cs:
```diff
@@ -0,0 +1,130 @@
+namespace TagsCloudVisualization.Tests
+{
+    [TestFixture]
+    public class CircularCloudLayouter_Tests
```
нижнее подчеркивание тут не нужно

skylog commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouter_Tests.cs:
```diff
@@ -0,0 +1,130 @@
+
+        
+        [TearDown]
+        public void RecordMistakes()
```
RecordMistakes - тоже не совсем корректное имя
- Record/Recording чаще используется для записи звука/видео, в задаче нужно вывести в консоль (можно использовать гл Print/Write)
- Mistakes - ошибки, но тут отлавливаем не ошибки, а падения тестов

skylog commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,79 @@
+
+        private Rectangle SearchRectangleOnCircle(Size rectangleSize)
+        {
+            while (true)
```
а без внешнего цикла не получилось сделать? нельзя ли по другому вызывать ResetRadius BoostRadius чтобы обойтись одним циклом?

skylog commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouter_Tests.cs:
```diff
@@ -0,0 +1,130 @@
+
+
+            PutManiRectangles(sizes);
+            Assert.IsFalse(AnyRectanglesIsIntersecting(cloudLayouter.GetRectangles()));
```
можно вызывать cloudLayouter.GetRectangles() внутри метода AnyRectanglesIsIntersecting
тогда в тестах будет: Assert.IsFalse(AnyRectanglesIsIntersecting())

skylog reviewed (commented):
ревью.

ivanarray committed:
Отрефакторил хелпер и тесты к нему

ivanarray committed:
Рефакторинг CircularCloudLayouter.cs и тесты на него
review: @skylog
#TagCloud

ivanarray committed:
рефакторинг CloudLayoutParameters.cs и тесты на него
review: @skylog
#TagCloud

ivanarray committed:
ренейм тестовых классов

ivanarray committed:
Merge remote-tracking branch 'origin/master'
